# <center>练习6</center>

<!-- toc -->

> #### 推荐时间 2天
>
> 第一天了解python的异常及try
>
> 第二天完成练习

## 练习目的

* 了解python的异常以及异常处理。
* 为什么需要异常处理

## 练习内容

以下练习在python的命令行模式完成即可

1. 尝试读取一个不存在的文件`notExistFile.json`，确认其异常类型
2. 复制以下代码到python命令行看会出现什么结果

    代码片段

    ``` python
    for i in range(10):
    print a
    ```

3. 尝试输入1+'a'，查看异常类型
4. `a=range(10)`，尝试获取`a[20]`，查看异常类型
5. 尝试`'1'.list`，了解其异常类型

以下练习内容写脚本完成

1. 尝试读取一个不存在的文件，文件不存在，脚本并不会退出，而是打印错误信息，该文件不存在
2. 为如下函数添加异常处理，使其在调用getValue(index=20)时不会出错退出，而是返回`None`

    代码片段

    ``` python
    def getValue(index=10):
	    mylist = ["Bob", "Steven", "Nancy", "Vicent", "Kevin"]
	    return mylist[index]
    ```

3. 以只读方式打开一个文件，尝试向其写入任意字符，捕获异常，使用raise提示错误，无论如何，在finally中将文件句柄关闭

## 练习要求

无

## 相关知识

* 为什么需要了解python的异常

    * 了解python异常之后，可以轻松的理解程序出错的原因（根据异常类型）
    * 可以防止一部分非法操作导致程序崩溃退出

* 什么时候需要异常处理呢？

    当python脚本执行的过程中，有时候会因为使用者用法错误，时常会发生异常，而如果不对异常进行处理，则会导致程序退出。如：

``` python
>>> f=file("test.txt")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IOError: [Errno 2] No such file or directory: 'test.txt'
>>> 1/0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ZeroDivisionError: integer division or modulo by zero
```


* 异常是什么？
	
    `python中一切皆对象`，异常也不例外。python提供了很多异常，当然也可以自定义异常。以下是一些常用的python异常：

异常名称 | 描述 | 常见于
------- |------- | --------
KeyboardInterrupt | 用户中断执行 | 通常是因为使用CTRL+C
AttributeError | 对象没有这个属性 | 使用了未定义的对象属性
EOFError | 没有内建输入,到达EOF 标记 | 使用CTRL+D可以导致EOF
IOError | 输入/输出操作失败 | 读取文件失败
ImportError | 导入模块/对象失败 | 没有这个模块或模块路径不正确
IndexError | 序列中没有此索引(index) | 超出了list的长度
KeyError | 映射中没有这个键 | 字典中无该键
SyntaxError | Python 语法错误 | 通常是缺少`:`，或引号，括号没有闭合等
IndentationError | 缩进错误 | 通常由于tab和空格混用导致
TypeError | 对类型无效的操作 | 状况较多，如1+“a”，函数参数个数不匹配等
NameError | 未声明/初始化对象 (没有属性) | 通常是该变量未初始化

* 异常处理

    * try...except...else

    ``` python
    try:
     	<语句>        #运行别的代码
    except <名字>：
        <语句>        #如果在try部份引发了'name'异常
    except <名字>，<数据>:
        <语句>        #如果引发了'name'异常，获得附加的数据
    else:
        <语句>        #如果没有异常发生
    ```

    * try...finally

    ``` python
    try:
    	<语句>
    finally:
    	<语句>    #退出try时总会执行
    ```

    * 触发异常

        可使用raise来指定异常类型

        ``` python
        raise [Exception [, args [, traceback]]]
        ```

        举例：
        ``` python
        def functionName( level ):
            if level < 1:
                raise Exception("Invalid level!", level)
            	# 触发异常后，后面的代码就不会再执行
        ```

## 思考

什么时候该用raise





